libs/beacon/src/beacon_manager.cpp

```diff
@@ -17,167 +17,617 @@
+void BeaconManager::AddCoefficients(MuddleAddress const &           from,
+                                    std::vector<Coefficient> const &coefficients)
 {
-  return contribution_.verification;
+  for (uint32_t ii = 0; ii <= polynomial_degree_; ++ii)
+  {
+    C_ik[identity_to_index_[from]][ii].setStr(coefficients[ii]);
+  }
 }
 
-BeaconManager::PrivateKey BeaconManager::GetShare(BeaconManager::Identity identity) const
+void BeaconManager::AddShares(MuddleAddress const &from, std::pair<Share, Share> const &shares)
 {
-  auto it = identity_to_index_.find(identity);
+  CabinetIndex from_index = identity_to_index_[from];
+  s_ij[from_index][cabinet_index_].setStr(shares.first);
+  sprime_ij[from_index][cabinet_index_].setStr(shares.second);
+}
 
-  if (it == identity_to_index_.end())
+/**
+ * Checks coefficients broadcasted by cabinet member c_i is consistent with the secret shares
+ * received from c_i. If false then add to complaints
+ *
+ * @return Set of muddle addresses of nodes we complain against
+ */
+std::unordered_set<BeaconManager::MuddleAddress> BeaconManager::ComputeComplaints()
+{
+  std::unordered_set<MuddleAddress> complaints_local;
+  for (auto &cab : identity_to_index_)
   {
-    throw std::runtime_error("Could not find identity in GetShare.");
+    CabinetIndex i = cab.second;
+    if (i != cabinet_index_)
+    {
+      // Can only require this if group_g_, group_h_ do not take the default values from clear()
+      if (C_ik[i][0] != zeroG2_ && s_ij[i][cabinet_index_] != zeroFr_)
+      {
+        bn::G2 rhs, lhs;
+        lhs = crypto::mcl::ComputeLHS(g__s_ij[i][cabinet_index_], group_g_, group_h_,
+                                      s_ij[i][cabinet_index_], sprime_ij[i][cabinet_index_]);
+        rhs = crypto::mcl::ComputeRHS(cabinet_index_, C_ik[i]);
+        if (lhs != rhs)
+        {
+          FETCH_LOG_WARN(LOGGING_NAME, "Node ", cabinet_index_,
+                         " received bad coefficients/shares from node ", i);
+          complaints_local.insert(cab.first);
+        }
+      }
+      else
+      {
+        FETCH_LOG_WARN(LOGGING_NAME, "Node ", cabinet_index_,
+                       " received vanishing coefficients/shares from ", i);
+        complaints_local.insert(cab.first);
+      }
+    }
+    ++i;
   }
+  return complaints_local;
+}
 
-  uint64_t n = it->second;
-  return contribution_.contributions[n];
+bool BeaconManager::VerifyComplaintAnswer(MuddleAddress const &from, ComplaintAnswer const &answer)
+{
+  CabinetIndex from_index = identity_to_index_[from];
+  if (identity_to_index_.find(answer.first) == identity_to_index_.end())
+  {
+    FETCH_LOG_WARN(LOGGING_NAME, "Node ", cabinet_index_,
+                   " received complaint answer with unknown reporter index");
+    return true;
+  }
+
+  CabinetIndex reporter_index = identity_to_index_[answer.first];
+  // Verify shares received
+  bn::Fr s, sprime;
+  bn::G2 lhsG, rhsG;
+  s.clear();
+  sprime.clear();
```

Comment:
I am confused about clear in this context - is it really neccessary to clear it immediately after constructing it? what does clearing it do?

Reply:
I believe the clear initialises the class members to zero. It may not be necessary but it is there as a precaution.